### PR TITLE
MQTT parser - v10

### PR DESCRIPTION
**Pull request by satta**
Previous PR: #5215  

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:  [#3549](https://redmine.openinfosecfoundation.org/issues/3549)

Describe changes (to previous PR):
- Pass existing `JsonBuilder` object to function instead of returning a new object.
- Remove occurrences of re-defining `MAX_SUBSTRINGS`.
- Add `if` checks for `NULL` pointers before calling `SCFree()` in MQTT related detect C code.
- Use `SCCalloc()` instead of plain `SCMalloc()` when allocating structs.
- Wrap `RegisterTests` assignments in UNITTESTS `#ifdef`s.

### CODE REVIEW

**Comment by satta** [comment]
macOS seems to have a build problem caused by crates.io download errors.

**Comment by victorjulien** [comment]
MacOS build failure looks like a runner failure.

**Comment by satta** [comment]
> MacOS build failure looks like a runner failure.

I think so too and am not going to do anything about it :)

File: rust/src/mqtt/detect.rs [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_unsubscribe_topic(tx: &MQTTTransaction,
+    i: u16,
+    buf: *mut *const u8,
+    len: *mut u32)
+    -> u8
+{
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::UNSUBSCRIBE(ref unsubv) = msg.op {
```
Comment (catenacyber) [inline rust/src/mqtt/detect.rs:393]: Can not there be more than one `UNSUBSCRIBE` in a transaction ?

Comment (catenacyber) [reply rust/src/mqtt/detect.rs:393]: This is my last question ;-)

**Comment by catenacyber** [comment]
> I think so too and am not going to do anything about it :)

I launched it again

File: rust/src/mqtt/detect.rs (hunk 1 above)
Comment (satta) [reply rust/src/mqtt/detect.rs:393]: Not according to the protocol, i.e. compliant brokers/clients would not create this. Forged or broken traffic might create such a situation if message IDs are used multiple times for separate UNSUBSCRIBE events.
I think I might really need the approach from https://github.com/OISF/suricata/pull/5208#discussion_r459466182 to make sure we get all of their topics if we want this function to work in this case as well (which I think we do!)

Comment (satta) [reply rust/src/mqtt/detect.rs:393]: See https://github.com/satta/suricata/commit/058c91b79cd43e587a49670eea0bd5e4580aa2ee

Comment (catenacyber) [reply rust/src/mqtt/detect.rs:393]: This commit looks good to me

Comment (satta) [reply rust/src/mqtt/detect.rs:393]: Cool, thanks for having a look. I will revisit and review the transaction code again to better define how the message ID ("packet identifier" in the spec) is used to group messages into transactions, in particular regarding ID reuse as addressed in the spec:

> The Packet Identifier becomes available for reuse after the Client has processed the corresponding acknowledgement packet. In the case of a QoS 1 PUBLISH this is the corresponding PUBACK; in the case of QoS 2 it is PUBCOMP. For SUBSCRIBE or UNSUBSCRIBE it is the corresponding SUBACK or UNSUBACK.

This means that we should make sure to never add to a 'completed' transaction where an acknowledgement has been seen. I'll make sure that is the case.

**Comment by satta** [comment]
Next PR: #5248

**Pull request closed without merging**